Feat: Apache ECharts

ArnavG-it opened the pull request:
![image](https://github.com/user-attachments/assets/89a3cdef-7c57-4c28-9c52-0f090f7f40a5)

The current Material UI line charts are not very easy to work with, or even designed for a heavier use case like Jarvis.

Apache ECharts provides lots of [customization](https://echarts.apache.org/en/option.html#title) and [examples](https://echarts.apache.org/examples/en/index.html), which is good for maintainability and flexibility.

This also moves the date range selection to the site header and provides it as global context for the components. That way, every component can refer to the same date range. The date picker is taken from [React Suite](https://rsuitejs.com/), which is a full UI library like Material UI.

Outstanding (doesn't have to be addressed in this PR):
- [ ] The tooltip/legend on the graphs is broken half the time (it will only label a few wells depending on where you hover) and I have no idea why. But more importantly, we should think of a proper design to communicate which line corresponds to which well since there are 16 of them.
- [ ] From my limited research, it's best to use just one of these large libraries (Suite and MUI) for a few reasons:
  - Performance and bundle size
  - Style conflicts (not just visually: they will try to overwrite each other's styles)
  - Style methodology. What I mean is that MUI encourages writing [wrappers to style components](https://github.com/UMSATS/Jarvis/blob/d9f040e522bd17533369cc26aa151634d152458d/client/src/styled-components/PageLayoutStyles.js#L9), which can use props and theme variables and is all done in JavaScript. Suite relies on traditional CSS (or LESS, which it recommends using) files and classes. Both can accomplish the same thing in the end, but again it's best to stick to one.

ArnavG-it committed:
Replace MUI charts with Apache Echarts

No luminosity graph for now.

ArnavG-it committed:
Global time range context

Allows both graphs to conform to the same time range, even if it is set on another page.

ArnavG-it committed:
Show inactive wells toggle

ArnavG-it committed:
Improve date range picker

ArnavG-it committed:
Filter x-axis instead of zoom

Allows using the zoom feature on the x-axis while still filtering based on the global time range. The zoom is not very smooth because it throttles its frame rate - the throttling can be turned off, but we'll see how real data performs first.

The tooltip is broken, but using axis indicators will be a better solution (16 wells in one tooltip is a lot).

ArnavG-it committed:
Add zoom reset button

ArnavG-it committed:
Use axis pointer instead of tooltip

ArnavG-it committed:
Use spread for common props

ArnavG-it commented:
@Koloss0 on Discord:

> Here are my thoughts:
> 
> - Thanks for looking into the date picker, this one is definitely more useable, but I'm not sure about the styling. Have you looked at anything like this one?: https://rsuitejs.com/components/date-range-picker/ I feel like something like that would fit with the theme a bit better.
> - I'm not totally in love with the idea of each graph having a dedicated zoom bar. Especially when we start having >3 graphs on the screen at once. It seems a bit overzealous to me. One idea is to have one on the top bar beside the date picker, and make it effect all the graphs at once. I'm not sure what would be the best way to go about that though.

The Suite one is definitely nice, with the buttons and default style. My main concern with it was that we have to install the entire React Suite package, which is on the scale of Material UI. That might affect performance when the site gets bigger, but I'm not sure. I don't mind trying it though. As for the current one, it can be styled fairly easily (quick example below), and the "OK" button could be manually added as well, but it is a bit more work.

![image](https://github.com/user-attachments/assets/f08e7e1d-dad6-488e-8bef-b3d728a02e30)

For the zoom bars, I added them in so you could get a feel for the zoom, but I agree that it's too much. There's also an option which lets you pan/zoom by just dragging/scrolling on the graph itself. It could also just be on the y-axis, since we have the date picker for the x-axis. Having it be a global thing requires using a different component for the bar since this one is built-in. For now, I might remove it, and we can revisit it later.

Koloss0 commented:
> My main concern with it was that we have to install the entire React Suite package, which is on the scale of Material UI. That might affect performance when the site gets bigger, but I'm not sure.

Oh, I didn't see that. That's fair. Hmmm, well this React Suite looks pretty nice. They also have a [TimePicker](https://rsuitejs.com/components/time-picker/), and TimeRangePicker, some [interesting stats](https://rsuitejs.com/components/stat/#stat-with-progress), and a [collapsible sidebar example](https://rsuitejs.com/components/container/#sidebar-layout). If we like the range picker, maybe we could think about switching over. The nice thing is it looks much closer to the design document out of the box.


> There's also an option which lets you pan/zoom by just dragging/scrolling on the graph itself.

I think I prefer scrolling being only for scrolling the page, otherwise it could get annoying.

> It could also just be on the y-axis, since we have the date picker for the x-axis. Having it be a global thing requires using a different component for the bar since this one is built-in. For now, I might remove it, and we can revisit it later.

Hmmm, I wonder if it would look good with a [slider](https://rsuitejs.com/components/slider/).

ArnavG-it commented:
> If we like the range picker, maybe we could think about switching over. The nice thing is it looks much closer to the design document out of the box.

We could do that. The theming/overriding is done differently than with MUI, but it should be easy to switch since we haven't done much styling yet. Think about it, and I'll add the date picker for now as a sample.

Edit: It didn't take long to encounter some conflicts. Material UI has [`<CSSBaseline>`](https://mui.com/material-ui/react-css-baseline/) which conflicts with some global theming that Suite's components require. I can work around it for this sample, but I think we should pick one or the other.

ArnavG-it committed:
Remove zoom and add legend tooltip

The legend design will need some consideration since we have a lot of wells/lines. The current one is temporary but could be used for a chart with just a few lines.

ArnavG-it committed:
Replace time range picker with React Suite version

I think I managed to do this without running into conflicts with Material UI. This is just using CSS, not LESS like they recommend in their docs, and I don't know if that locks us out of any customization. Works fine for this component though.

ArnavG-it committed:
Improve context usage

ArnavG-it committed:
Add disabled date ranges, minor styling

ArnavG-it commented:
@Koloss0 This is ready for review/feedback, whenever you get the chance. In the main post I mentioned a bug with the legend (you'll see it if you refresh the page a few times). I could remove the legend for now if you'd rather not push a bug. Also, I mentioned some considerations about React Suite. I think it's mostly fine for now, although the date picker calendar does take a second to close which could be due to those issues (it still lags when the component has no hooks).

Oh, and the graph titles are a bit ugly by default. Let me know if you have a design in mind, or we can come back to it later.

ArnavG-it committed:
Minor graph styling

Disables pointer cursor and left padding

Koloss0 reviewed (approved):
I think it's fine for now. We can discuss these topics again in today's meeting.
